### ISSUE TITLE
Possibility of enabling casttype for repeated

### ISSUE THREAD

**Issue by silasdavis** [gogo/protobuf#433]
Often I define named types like:

```go
type Balances []Balance

func (b Balances) DoSpecialStuff() {
...
}
```
Where `Balance` may be a protobuf type it would be useful to be able to generate the named type for the slice which should be castable with something like:

```protobuf
repeated Balance Balances = 1 [(gogoproto.repeatedcasttype) = "Balances"];
```

Worth it? Doable?

**Comment by awalterschulze** [comment]
Sorry for the delay, this is very interesting.

I can think of some reasons, but may I ask why you like using this pattern?

**Comment by silasdavis** [comment]
There are a number of reasons I tend to use it:

1. To control the rendering/marshalling of a value, e.g. `type HexInts []uint64` where I can define a MarshalJSON (or a codec by some other means) and have the value rendered as fixed-width hex for serialisation - if that is more appropriate for some reason.

2. To provide an implementation of `sort.Interface` so that my collection is sortable 

3. To provide helpers on the collection preserves certain properties of the collection or provide a more convenient mutation interface

The main reason for not just using a struct wrapper is the same as for sometimes preferring a named type in general, but in particular you can still use range interation (i.e. `i,b := range Balances`) without a cast as well as benefit from an implicit cast to the named type when calling functions with a `Balances` argument and returning arguments of `Balances` type.

**Comment by awalterschulze** [comment]
I only thought of number 2, very interesting.

@donaldgraham @jmarais 

What do you think?

**Comment by donaldgraham** [comment]
This sound great.  @jmarais, let's give some more thought to this use case.

**Comment by GeertJohan** [comment]
I'd like to add a possible use-case/example for this.

I'm using the generated messages directly when scanning values from `database/sql`. Sometimes via [sqlx's StructScan](https://godoc.org/github.com/jmoiron/sqlx#Rows.StructScan). This allows for extremely simple code, some API's are implemented in a few minutes and perform very well.

When a column in the postgres database result rows is of type `[]text`, it can be scanned using a Go type that knows how to handle the bytes from the protocol. For postgres this is [pq.StringArray](https://godoc.org/github.com/lib/pq#StringArray), which is defined as `type StringArray []string`. Sounds like the perfect solution! However, I cannot use pq's StringArray as casttype for `repeated string`, because that will (correctly) result in generated code `[]pq.StringArray`. I need it to be just `pq.StringArray`.

Perhaps there could be an option castslicetype, which is only allowed on repeated fields and casts the complete slice, not the values _in_ the slice.

**Comment by rohanil** [comment]
any updates here?

**Comment by ivanovaleksey** [comment]
@GeertJohan I use it the same pattern. Do you have any solution for working with `pq.StringArray`?

**Comment by GeertJohan** [comment]
@ivanovaleksey In this scenario I use a separate struct for proto and sql, and copy over the values.

**Comment by ivanovaleksey** [comment]
@GeertJohan thanks for the reply! I thought there might be a solution with embedding and overriding field type, but I can't manage to work it for now.

**Comment by ivanovaleksey** [comment]
@GeertJohan I managed to get working solution with embedding. Maybe it would be useful for you or someone else.
```
// proto
message Whatever  {
...
repeated string tags = 7 [(gogoproto.moretags) = "db:\"-\""];
}

// go
var obj struct {
	Whatever
	Tags pq.StringArray `db:"tags"`
}
err := db.QueryRowx(query).StructScan(&obj)
```
As I understand `db: "-"` was an important part.

**Comment by GeertJohan** [comment]
Ahter the db scan, you do still have to copy from obj.Tags to Whatever.Tags, right?

**Comment by ivanovaleksey** [comment]
@GeertJohan yes, you are right. But copying 1 line is better in my case than converting a whole db struct into proto struct (one though can create helper method to hide complexity)
